### ISSUE TITLE
make `impl_has_work` compatible with more complex generics

### ISSUE THREAD

**Issue by BennoLossin** [Rust-for-Linux/linux#1077]
The `impl_has_work` macro only supports `ident`s as generic arguments:

```rust
macro_rules! impl_has_work {
    ($(impl$(<$($implarg:ident),*>)?
       HasWork<$work_type:ty $(, $id:tt)?>
       for $self:ident $(<$($selfarg:ident),*>)?
       { self.$field:ident }
    )*) => {$(
```

There is a workaround, see Alice's [linked list patch](https://lore.kernel.org/rust-for-linux/20240402-linked-list-v1-4-b1c59ba7ae3b@google.com/):
```rust
    macro_rules! impl_list_item {
        (
            impl$({$($generics:tt)*})? ListItem<$num:tt> for $t:ty {
                using ListLinks;
            } $($rest:tt)*
        ) => {
```

Make the `impl_has_work` macro compatible with more complex generics such as lifetimes and const generic arguments.

---
This requires submitting a proper patch to the LKML and the Rust for Linux mailing list. Please recall to test your changes, to use a proper title for the commit, to sign your commit under the Developer's Certificate of Origin and so on. Please see https://docs.kernel.org/process/submitting-patches.html and https://rust-for-linux.com/contributing for details.

Please take this issue only if you are new to the kernel development process and you would like to use it as a test to submit your first patch to the kernel. Please do not take it if you do not plan to make other contributions to the kernel.

**Comment by mu001999** [comment]
I'd like to try this, and I sent an email, https://lkml.org/lkml/2024/5/22/651. I don't know if I did the right thing

**Comment by ojeda** [comment]
fe7d9d804337, which is in linux-next, adds the feature. Thanks!

**Comment by Darksonn** [comment]
This was merged in rust-6.11.
